File: LAB 2/MIDPOINT_CIRCLE.m
clc; clear; close all;

RADIUS = 10;
CX = 20;
CY = 20;

figure('Position', [100 100 600 500]);
hold on;
grid on;
axis equal;
axis([CX-RADIUS-5 CX+RADIUS+5 CY-RADIUS-5 CY+RADIUS+5]);
title('Midpoint Circle Algorithm', 'FontSize', 14);
xlabel('X-axis', 'FontSize', 12);
ylabel('Y-axis', 'FontSize', 12);


x = 0;
y = RADIUS;
d = 1 - RADIUS;


points_x = [];
points_y = [];


while x <= y

    sym_points = [x, y; -x, y; x, -y; -x, -y;
                  y, x; -y, x; y, -x; -y, -x];


    actual_x = CX + sym_points(:,1);
    actual_y = CY + sym_points(:,2);


    points_x = [points_x; actual_x];
    points_y = [points_y; actual_y];


    if d < 0
        d = d + 2*x + 3;
    else
        d = d + 2*(x-y) + 5;
        y = y - 1;
    end
    x = x + 1;
end


plot(points_x, points_y, 'ks', 'MarkerSize', 8, 'MarkerFaceColor', 'k');


theta = 0:0.01:2*pi;
circle_x = CX + RADIUS * cos(theta);
circle_y = CY + RADIUS * sin(theta);
plot(circle_x, circle_y, 'b-', 'LineWidth', 2);


plot(CX, CY, 'ro', 'MarkerSize', 12, 'MarkerFaceColor', 'r');


text(CX, CY+RADIUS+3, sprintf('Radius = %d', RADIUS), ...
    'HorizontalAlignment', 'center', 'FontSize', 10, 'FontWeight', 'bold');

File: LAB 2/Python/MIDPOINT_CIRCLE.py
ximport matplotlib.pyplot as plt
import numpy as np

RADIUS = 8
CX, CY = 10, 10

plt.figure()
plt.grid(True)
plt.axis('equal')
plt.title('Midpoint Circle Algorithm')
plt.xlabel('X-axis')
plt.ylabel('Y-axis')

X, Y = 0, RADIUS
P = 1 - RADIUS

X_VALS, Y_VALS = [], []

def plot_symmetric_points(X_VALS, Y_VALS, CX, CY, X, Y):
    points = [(X, Y), (-X, Y), (X, -Y), (-X, -Y),
              (Y, X), (-Y, X), (Y, -X), (-Y, -X)]
    for PX, PY in points:
        X_VALS.append(CX + PX)
        Y_VALS.append(CY + PY)
    return X_VALS, Y_VALS

X_VALS, Y_VALS = plot_symmetric_points(X_VALS, Y_VALS, CX, CY, X, Y)

while X < Y:
    X += 1
    if P < 0:
        P = P + 2 * X + 1
    else:
        Y -= 1
        P = P + 2 * (X - Y) + 1
    X_VALS, Y_VALS = plot_symmetric_points(X_VALS, Y_VALS, CX, CY, X, Y)

THETA = np.linspace(0, 2*np.pi, 100)
X_CIRCLE = CX + RADIUS * np.cos(THETA)
Y_CIRCLE = CY + RADIUS * np.sin(THETA)

plt.plot(X_CIRCLE, Y_CIRCLE, 'b-', linewidth=1.5)
plt.plot(X_VALS, Y_VALS, 'ks', markersize=8, markerfacecolor='k')
plt.plot(CX, CY, 'ro', markersize=10, markerfacecolor='r')
plt.show()
